File: Chapter7Search/DSC++/Node.cpp
#pragma once
#include "Edge.cpp"
#include <string>
#include <array>
class Node {
private:

	std::string value;
	std::array<Edge, 5> *leftLink;

	int indexOf(Edge edge) {
		for (int i = 0; i < leftLink->size; i++)
		{
			if (leftLink->at(i) == edge) {
				return i;
			}
		}
		return -1;
	}

public:

	Node() {
		value = "Node";
	}

	Node(std::string value) {
		this->value = value;
	}

	bool addEdge(long value, Node next) {
		Edge edge(value, next);
		if (this->indexOf(edge) >= 0) {
			return false;
		}
		else {
			leftLink->assign(edge);
			return true;
		}
	}

	int leftLinkSize() {
		return leftLink->size();
	}

	Edge getEdge(int index) {
		if (index > leftLink->size() - 1) {
			return;
		}
		return leftLink->at(index);
	}

	std::string getValue() {
		return value;
	}

	//thx "stack overflow"
	bool operator==(const Node& n) const {
		return (value == n.value) && (leftLink == n.leftLink);
	}

	~Node() {
		delete leftLink;
	}
};

File: Chapter7Search/DSJava/Node.java
package com.company;

import java.util.ArrayList;

public class Node {
    private Object value;
    private ArrayList<Edge> leftLink;

    public Node(Object value) {
        this.value = value;
        this.leftLink = new ArrayList<>();
    }

    public boolean addEdge(long value, Node next) {
        Edge edge = new Edge(value,next);
        if (leftLink.contains(edge)) {
            return false;
        } else {
            leftLink.add(edge);
            return true;
        }
    }

    public int leftLinkSize(){
        return leftLink.size();
    }

    public Edge getEdge(int index){
        if(index>leftLink.size()-1){
            return null;
        }
        return leftLink.get(index);
    }

    public ArrayList<Node> getAllNeighbours() {
        ArrayList<Node> neighbours = new ArrayList<>();

        for (Edge e:this.leftLink) {
            neighbours.add(e.getLeftLink());
        }

        return neighbours;
    }

    public Object getValue() {
        return value;
    }
}
